Add memc support for STM32F4 series

fabiobaltieri opened the pull request:
Hi, this adds FMC support to STM32F4 series SoC. The current driver is only enabled for some H7 series, but seems to work just fine on F4 as well. this adds the device node to stm32f427.dtsi (sourced by the f429 as well) and enables it for stm32f429i_disc1, which ships with an 8MB SDRAM chip.

For the discovery board, the timings are copied from https://github.com/STMicroelectronics/STM32CubeF4/blob/master/Projects/STM32F429I-Discovery/Examples/FMC/FMC_SDRAM/Src/main.c#L89-L114, though I had to use CAS_2 for some reason. Tested using tests/drivers/memc/stm32_sdram (modified to cover the whole 8MB), seems to work fine.

fabiobaltieri committed:
dts: stm32f4: add fmc device node for f4 series devices

Add a device node for the FMC controller found in stm32f427, f437 and
f446 devices, works fine with the current memc_stm32_sdram driver.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri committed:
boards: stm32f429i_disc1: add sdram support

Add support for the 8MB SDRAM found in STM32F429-DISC1 boards.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

erwango commented on dts/arm/st/f4/stm32f427.dtsi:
```diff
@@ -86,5 +87,21 @@
 			label = "SPI_6";
 		};
+
+		fmc: memory-controller@52004000 {
```
Would you mind adding nodes as well for f437 and f446 ?
This would make the support complete for the F4 family.

fabiobaltieri replied:
Yeah no problem. Fixed the node address as well, realized now it was not matching the reg property.

erwango reviewed (changes requested):
Thanks for this work!

Looks good, but I would have expected to have an update also on the following line:
https://github.com/zephyrproject-rtos/zephyr/blob/c9aa260f0cfc54453aef70f876e9ba9385428a30/drivers/memc/Kconfig.stm32#L6
Can you have a check ?

Also, but while we're at it, it would be great to provide the complete series support.

fabiobaltieri committed:
drivers: stm32-fmc: enable MEMC_STM32 default based on dt

Default the option MEMC_STM32 to "y" when the device node is defined and
enabled, so that the driver is selected automatically when the board
supports it and MEMC is enabled.

Remove the default conditional on serise as it's redundant with the one
in soc/arm/st_stm32/common/Kconfig.defconfig.series.

Signed-off-by: Fabio Baltieri <fabio.baltieri@gmail.com>

fabiobaltieri commented:
> Thanks for this work!
> 
> Looks good, but I would have expected to have an update also on the following line:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/c9aa260f0cfc54453aef70f876e9ba9385428a30/drivers/memc/Kconfig.stm32#L6
> 
> Can you have a check ?

Yeah, good point. I see what happened here, there's an extra "MEMC_STM32 depends on MEMC" in soc/arm/st_stm32/common/Kconfig.defconfig.series. So the option overall was looking like this:

```
Direct dependencies (=y):
     MEMC(=y) && SOC_FAMILY_STM32(=y)  (=y)
  || SOC_SERIES_STM32H7X(=n) && MEMC(=y)  (=n)
```

Making the memc one ineffective.

I added a patch to remove that series level depend, and made the option depend on the device node instead. Thinking it may be convenient to have it "default y" as well, but then the device would always be configured for the board and it'd get in the way of power measurement tests. Let me know what you think.

> 
> Also, but while we're at it, it would be great to provide the complete series support.

gmarull commented on soc/arm/st_stm32/common/Kconfig.defconfig.series:
```diff
@@ -71,8 +71,4 @@ config DMA_STM32
 
-config MEMC_STM32
-	default y
-	depends on MEMC
```
All other peripherals do the same (see the lines above). This enables MEMC_STM32 automatically when MEMC is enabled, so I think it does its job.

erwango replied:
Indeed, you should keep this line and remove `CONFIG_MEMC_STM32=y` in `tests/drivers/memc/stm32_sdram/prj.conf `

fabiobaltieri replied:
Don't think the other peripheral settings are right either.

So with that block in Kconfig.defconfig.series, MEMC can be selected for every STM32 part (current code) I can, say, select an F103 board, enable MEMC, enable MEMC_STM32 and get a config that does not build cause there's no fmc node.

With that code removed, MEMC_STM32 can only be selected if MEMC is selected (the Kconfig is not included if not) and the node is present (does not build otherwise anyway),

If we want to get rid of the line in prj.conf, I can switch the MEMC (which depends on the node) to default y, in that case it can only be selected if the node is defined and it also does not have to be selected explicitly, and will just go in if MEMC is enabled. I'll do that change, lmk what you think.

fabiobaltieri replied:
The other options should be reviewed as well, for example I can do

`west build -p -b nucleo_wl55jc samples/basic/blinky`

and then in menuconfig I can enable `USB` and `USB_DC_STM32` gets enabled automatically (the device does not have an USB controller), resulting in a broken build.

erwango replied:
ok, right with the compat enabled, this should indeed be removed. Txs

fabiobaltieri commented:
Rebased, dropped the explicit `CONFIG_MEMC_STM32` from the test `prj.conf` file, changed the `MEMC_STM32` default to yes.

Tested:
- build against `stm32f429i_disc1`: selecting `MEMC` enables the driver
- build against `stm32_min_dev_blue`: selecting `MEMC`, the FMC option is not available

erwango commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1.dts:
```diff
@@ -109,3 +114,42 @@
+					    STM32_FMC_SDRAM_RPIPE_1>;
+			st,sdram-timing = <2 7 4 7 2 2 2>;
+		};
+
```
Nit: remove blank line

fabiobaltieri replied:
Oh yeah, dropped.

erwango reviewed (changes requested):
One minor comment, otherwise LGTM

gmarull commented on drivers/memc/Kconfig.stm32:
```diff
@@ -1,9 +1,12 @@
 config MEMC_STM32
 	bool "Enable STM32 Flexible Memory Controller (FMC)"
-	depends on SOC_SERIES_STM32H7X
+	depends on $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_FMC))
+	default y
```
`default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_FMC))`

fabiobaltieri replied:
So I'm dropping the `depends` and using the compat for `default`. Fine by me, though then it allows selecting the option without the node being present.

fabiobaltieri replied:
...it also allows selecting the driver on any architecture, let me know if I misinterpreted the suggested change.

gmarull replied:
I think that Kconfig needs to specify software dependencies, that is, the driver implementation depends on H7 and F4 series not on a node enabled in DT. With current solution, one can also define the FMC node in any series and get it compiled. So I guess the right way would be:

```
depends on SOC_SERIES_STM32H7X || SOC_SERIES_STM32F4X
default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_FMC))
```

fabiobaltieri replied:
Ok changing it to that. Thing is that those lists are getting out of control for some of the other drivers. Thinking that enabling these from the soc with a series of `HAS_STM32_XXX` options may be cleaner. Anyway, not in scope for this PR so let's go with your suggestion.

gmarull replied:
looks like e49c849 commit message needs to be adjusted now

fabiobaltieri replied:
Right, updated.

fabiobaltieri commented:
Updated the `MEMC_STM32` conditions.

erwango commented on drivers/memc/Kconfig.stm32:
```diff
@@ -1,9 +1,12 @@
 config MEMC_STM32
 	bool "Enable STM32 Flexible Memory Controller (FMC)"
-	depends on SOC_SERIES_STM32H7X
+	depends on SOC_SERIES_STM32H7X || SOC_SERIES_STM32F4X
```
Seems we're back to square one.
These 2 lines should be redundant. Did you face a specific issue ?

fabiobaltieri replied:
Well I still removed the block in `Kconfig.defconfig.series`, so these are not redundant anymore now and kind of make sense. The only issue I see now is maintaing this list (likely not a big deal for MEMC specifically).

erwango replied:
Well, the use of compatible macro here is to be able to use device tree as the main source of HW support.
So I prefer we keep the line in `soc/arm/st_stm32/common/Kconfig.defconfig.series` and remove this list.
Also this consistent with what we do for (most of) other drivers.

gmarull replied:
Most drivers have, in fact, `depends on SOC_FAMILY_STM32`

fabiobaltieri replied:
I liked the `depends on $(dt_...)` cause it's representative of the fact that the driver code itself depends on the node being defined and enabled, so that has the lowest chances of resulting in configuration that does not build. Are you aware of a specific Zephyr wide guideline to follow or are we just trying to be coherent?

erwango replied:
> Are you aware of a specific Zephyr wide guideline to follow or are we just trying to be coherent?

In general, we're trying to minimize use of Kconfig when possible and use device tree as main source of hw support when possible.

 > Most drivers have, in fact, depends on SOC_FAMILY_STM32, but that will show the option for series that do not have FMC available.

I agree this is not ideal. Maybe I'm minimizing the impact because I'm not a major menuconfig user, but in another hand I haven't seen a lot of complains on this particular topic.
Anyway, we should provide some better solution, but this should rely on device tree rather than Kconfig, as with Kconfig we should anyway maintain a list of SERIES/SOC symbols, which I don't think is reasonable.
In the end, using device tree, we should be able to both detect:
- what node is available
- what node is enabled

fabiobaltieri replied:
Sounds like

```
depends on $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_FMC))
default y
```

to me. That's what I'd like anyway. :-)

erwango replied:
> depends on $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_FMC))
> default y

IIC that's equivalent to `default y $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_FMC))`.
Also, can you revert the lines in `soc/arm/st_stm32/common/Kconfig.defconfig.series` ?

Then we'll be on par with existing practice in other drivers and we'll be able to merge.

About controlling more strictly drivers symbols, I'm not opposed to investigate and do some clean up.
But let's see if this can be addressed globally (at least on STM32 level), identifying exactly the issue, the impact, and how we can use dts to fix that.

fabiobaltieri replied:
Yeah let's address it separately, so I reverted the `soc/arm/st_stm32/common/Kconfig.defconfig.series`, leaving these are they are now. LMK if that's what you mean.

erwango replied:
Let's alos 

> Yeah let's address it separately, so I reverted the `soc/arm/st_stm32/common/Kconfig.defconfig.series`, leaving these are they are now.

This and then we miss removing the following:
```
	depends on SOC_SERIES_STM32H7X || SOC_SERIES_STM32F4X
```

fabiobaltieri replied:
Alright, done.

gmarull reviewed (approved):
Thanks!
